core/iwasm/interpreter/wasm_interp_classic.c

```diff
@@ -3716,7 +3716,9 @@ wasm_interp_call_func_bytecode(WASMModuleInstance *module,
                           value: readv * 0 + sval */
                         DEF_ATOMIC_RMW_OPCODE(XCHG, *0 +);
                 }
-
+#if WASM_ENABLE_THREAD_MGR != 0
+                CHECK_SUSPEND_FLAGS();
```

Comment:
Adding `CHECK_SUSPEND_FLAGS` here impacts each opcode of shared memory (atomic load/store, atomic wait/notify, rmw cmpxchg, etc.), it may greatly degrade the performance.

My understanding is that when main thread encounters proc_exit, it will throw and spread the exception, and notify all the atomic waiting nodes of other thread, and when the other thread receives the notification, it doesn't check the exception again, and runs into unreachable opcode, and overwrites the original proc_exit exception with unreachable exception.

We may add extra checks instead, e.g., in the handling of atomic.wait32/wait64 in interpreter:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_interp_fast.c#L3266-L3267
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_interp_fast.c#L3286-L3287
Change
```C
    if (ret == (uint32)-1)
        goto got_exception;
```
to
```C
    if (ret == (uint32)-1 || wasm_get_exception(module))
        goto got_exception;
```
Or directly modify wasm_runtime_atomic_notify:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/common/wasm_shared_memory.c#L482
Change `return is_timeout ? 2 : 0;` to
```C
if (wasm_runtime_get_exception(module))
    return -1;
return is_timeout ? 2 : 0;
```

Reply:
When the other thread receives the notification, we just want to check if the exception flag was set and, if that's the case, we want to return instead of fetching the next instruction using `HANDLE_OP_END`.
As you were pointing out, putting `CHECK_SUSPEND_FLAGS()` after each atomic instruction is too much overhead, we can just move it into the atomic wait opcode handling.

About `unreachable`, it just happened to be the next instruction in my test case
```wasm
i32.const 0
i32.const 0
i64.const -1
memory.atomic.wait32
unreachable
```

Reply:
OK, had you tested the AOT mode? Seems that we should also add the suspend flag check after calling wasm_runtime_atomic_wait in AOT code:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot_emit_memory.c#L1370

Refer to:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/compilation/aot_emit_control.c#L753-L759

Reply:
Added a commit for that, but it doesn't work. Any suggestion for debugging AOT in general?

Reply:
Not a simple way to debug the AOT code now, normally we step into the assembler instructions to debug it.

Did you add `--enable-multi-thread` for wamrc to generate the AOT file?

Reply:
Do I need it? I'm testing using this sample but I replace the busy waiting with a `__builtin_wasm_memory_atomic_wait32(0, 0, -1)`
https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/samples/wasi-threads/wasm-apps/thread_termination.c#L34

Without `--enable-multi-thread`, I see that it exits the wait but fetches the instruction after (i.e. what I want to avoid with this PR). The next instruction being this assert https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/samples/wasi-threads/wasm-apps/thread_termination.c#L43

With `--enable-multi-thread` it crashes here https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/common/wasm_runtime_common.c#L197
But I don't think I need `--enable-multi-thread`, at least I wasn't using it with WASI threads in the past.

Reply:
Actually, I see problems in general when using AOT: this line
https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/samples/wasi-threads/wasm-apps/no_pthread.c#L65
never gets executed. After some debugging, it looks like, after the atomic wait, the main thread ends and start waiting for the spawn thread https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/common/wasm_application.c#L218 without executing the instructions after the wait.

All this happens without the changes in `core/iwasm/compilation/aot_emit_memory.c` that I added with this PR.
Not sure how long the problem has been there.

Reply:
I think I found the problem, the atomic wait returns `-1` in case of error, instead here we check for `> 0`.
https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/compilation/aot_emit_memory.c#L1347
I quickly fixed it by using `LLVMIntSGE` instead of `LLVMIntSGT` and it seems to work. Now I need to do it properly by using a cast like here https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/interpreter/wasm_interp_classic.c#L3422

Reply:
Can use `BUILD_ICMP(LLVMIntNE, ret_value, I32_NEG_ONE, cmp, "atomic_wait_ret");` for `if (ret == (uint32)-1`

Reply:
> https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/samples/wasi-threads/wasm-apps/thread_termination.c#L43
> 
> With `--enable-multi-thread` it crashes here
> 
> https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/common/wasm_runtime_common.c#L197
> 
The exception should have been set before accessing the guard page:
[wasm_runtime_access_exce_check_guard_page](https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/common/wasm_runtime_common.c#L1584)
Not sure whether it was cleared by `clear_wasi_proc_exit_exception`?

> 
> But I don't think I need `--enable-multi-thread`, at least I wasn't using it with WASI threads in the past.
> [EDIT] It is failing already at compilation time, `if (comp_ctx->enable_thread_mgr)` evaluates to false (without `--enable-multi-thread`), I'm debugging that.

It should be set, or `if (comp_ctx->enable_thread_mgr)` will be false.

Reply:
> The exception should have been set before accessing the guard page:
[wasm_runtime_access_exce_check_guard_page](https://github.com/bytecodealliance/wasm-micro-runtime/blob/92c4bbebadb1589bda96d4365300723e06082dde/core/iwasm/common/wasm_runtime_common.c#L1584)
Not sure whether it was cleared by clear_wasi_proc_exit_exception?

Yes, that's exactly what's happening. Depending on the run, sometimes the exception gets cleared before and the assert makes it crash. Sometimes it doesn't and the behavior is correct.

Reply:
Also, what's the equivalent of https://github.com/bytecodealliance/wasm-micro-runtime/blob/63273a1673d27df64159c0511f30e3047806a13b/core/iwasm/interpreter/wasm_interp_classic.c#L3821 in AOT mode? Because I need to check for termination there as well.

Reply:
I removed the assert and added the termination check in `aot_emit_function.c`. I ran a few test programs and it seems to work.

Reply:
Ok, the problem was that we don't set the suspend flag for the thread that originally raises the exception.